Add ppc64le & multiarch support

sdmonov opened the pull request:
This code is adding ppc64le and multiarch support for FfDL.

- Adds new make target to the main Makefile: **docker-create-manifest**
  that creates the multiarch manifests for the services.

- arch (architecture) parameter was added to values.yaml and
  storage-plugin/values.yaml

**make docker-build** will generate all the services with a -${ARCHITECTURE}
suffix. During build an ARCH argument is sent to the Dockerfiles in order
to have architecture specific implementation inside it.

Building docker images for ppc64le should look like this:
```
make docker-build
make docker-push
make docker-create-manifest
.....
helm install storage-plugin --set arch=ppc64le,namespace=$NAMESPACE
helm install . --set arch=ppc64le,lcm.shared_volume_storage_class=$SHARED_VOLUME_STORAGE_CLASS,namespace=$NAMESPACE

```
Fixes #115

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

whummer committed:
Merge pull request #2 from whummer/fix/travis-img

[don't merge yet] Fix package names and travis config

animeshsingh committed:
Update README.md (#4)

animeshsingh committed:
Adding paper references - changing main title (#5)

animeshsingh committed:
Removing the extra bracket around references (#7)

Tomcli committed:
Update instructions with Cloud Object Storage (#3)

Update instructions with Cloud Object Storage

animeshsingh committed:
Architecture diagram + adjusted specs for memory and cpu (#12)

* arch diag

* arch diag

* Update README.md

* adding specs

* adding specs

Tomcli committed:
Add support to none VM_TYPE Clusters (#15)

* update none VM_TYPE

* polish export commands

Tomcli committed:
add user guide (#10)

Tomcli committed:
* Fix Caffe Example (#19)

* Extent make test-submit waiting time

Tomcli committed:
Assign default edit role to lcm to support kubernetes 1.9 and above (#22)

* assign default edit role to lcm

* add helm value options for 1.7 and below

animeshsingh committed:
Some changes in Prereqs + Pointers to ICP  (#23)

* adding prereqs and bumping user guide in front

* adding prereqs and bumping user guide in front

Tomcli committed:
Don`t Merge Yet - Add Caffe2, PyTorch and GPU support(using Accelerators) (#24)

* add caffe2 and pytorch cpu support

* update LCM, learner config file, and example jobs

* fix pytorch example bug

* Update gpu-guide.md

* Update gpu-guide.md

* merge CPU and GPU examples into a single example

* add more tf framework versions

* fix typo

* add S3 prereq

Tomcli committed:
Update UI instructions to manually create model zip files (#29)

* update UI instructions

* fix command

animeshsingh committed:
revised architecture details (#30)

Tomcli committed:
Update helmignore to avoid oversized helm configmap (#32)

Tomcli committed:
update UI dependencies and add font-awesome (#33)

whummer committed:
add initial list of contributors (#36)

animeshsingh committed:
Adding Contributors in CONTRIBUTING.MD (#37)

* adding contributors

* Update README.md

animeshsingh committed:
Updating maintainers file (#38)

* Updating maintainers file

* Update MAINTAINERS.md

Tomcli committed:
Add converting script for FfDL to WML and vice versa (#39)

* add converting script

* update converter readme and update tensorflow version

* update troubleshooting

* Update README.md

* Update gpu-guide.md

* Update README.md

* Update README.md

* Update README.md

animeshsingh committed:
Linkage to watson studio  (#40)

* Adding references to Watson Studio

* Update README.md

* Rename README.md to ffdl-wml.md

* Update README.md

* Create train-deploy-wml.md

* Update train-deploy-wml.md

* Update README.md

* Update ffdl-wml.md

* Update ffdl-wml.md

* Update ffdl-wml.md

* Update README.md

* Update README.md

* update WML instructions

* revert tf example

* update caffe manifest

Tomcli committed:
Remove Vagrant reference (#43)

Tomcli committed:
update tensorflow example (#44)

Tomcli committed:
Temporary fix for CI with k8s 1.9.4 (#45)

* Update feature-gates for k8s 1.9.4 and above

* Update troubleshooting

* Update README.md

Tomcli committed:
Update learner entrypoint command to work on k8s 1.9.4 and above (#47)

* update learner entrypoint command

* update troubleshooting

* Update train.sh path to match with DLaaS.

animeshsingh committed:
Removing orphan vagrant references (#50)

Tomcli committed:
Add developer guide and some customizable features for development build (#51)

* * Add codebase configuration for device plugin and custom learner images
* Add developer guide for those who want to do a custom FfDL build

* update developer-guide

* fix declare type

kant committed:
Minor fix on #273 & #370 (#35)

Plus minor fixes

animeshsingh committed:
Creating CLA (#52)

* Creating CLA

* Update CLA.md

* Update CONTRIBUTING.md

Tomcli committed:
fix logging bug (#53)

lresende committed:
Move DCO signoff as part of the PR Template (#58)

lresende committed:
Update LICENSE and generate NOTICE file (#59)

Some dependencies used in FfDL and included as source
are not licensed as AL2 and thus we should add them
explicitly in the main LICENSE file. Other dependencies
require attribution, and thus mention it on NOTICE file.

Tomcli committed:
fix dependencies issue (#63)

* fix dependencies issue

* update prereq

animeshsingh committed:
Compiling a list of Models we can train - expect this to grow (#66)

* sample models

* Update models.md

* Update models.md

animeshsingh committed:
Design details (#67)

* design-details

* design-details

* design-details

Tomcli committed:
update example resources (#68)

LGTM

animeshsingh committed:
Collaborators - File for summarizing collaborators (#69)

* design-details

* design-details

* design-details

* adding collaborators

Commit:
Add missing Python imports (#70)

Thanks for the PR @cclauss. 
@Tomcli thanks for review

Tomcli committed:
* fix ui dependencies and temperary fix helm issue in travis (#73)

* Clarify Minikube's develop environment

Tomcli committed:
add support for latest tf framework versions (#75)

Tomcli committed:
Update python3 dependencies for ubuntu 16 (#76)

* Update python3 dependency for ubuntu 16

* Update python3 dependencies

* Update Dockerfile

* Fix dependencies on emetrics image

* Update Dockerfile

Tomcli committed:
Revert "Update python3 dependencies for ubuntu 16 (#76)" (#80)

This reverts commit a18bde590b6d65afa2f8aa9b71bac84ed673351d.

ukclivecox committed:
Update tf mnist model to output class probabilities via softmax (#83)

ukclivecox committed:
Add Seldon Deployment Example for tf-model (#86)

* initial commit for seldon-ffdl integration

* update docs and add example serving tests

* clean notebooks

* Update README.md

* fix typos

* Updates from review comments - bucket secrets additions, bucket name env var, typos

* clean serving notebook

ukclivecox committed:
Fix typo in TFMnist URL link (#87)

nkpng2k committed:
PR - for FfDL Integration With H2O.ai (#88)

* adding changes for h2o3 integration

* minor edits

* reordered swith case in container_helper_extensions.go

* commit to resolve multinode failures issue

* adding changes for h2o3 integration

* minor edits

* reordered swith case in container_helper_extensions.go

* commit to resolve multinode failures issue

* adding collaborator files inside community folder

* adding simple readme

* fixing repo

* reverting edits to Makefile

* additional instructions to obtain toy dataset

* increasing default resources in manifest-h2o.yml

* Edit README.md with minimum Memory suggestions

Tomcli committed:
Clean up and rearrange training instructions (#89)

* clean up and rearrange training instructions

* Update README.md

ukclivecox committed:
Add Seldon PyTorch deployment example (#91)

* add seldon pytorch example

* Review fixes - typos and remove jupyter checkpoints folder

* Update memory requirements for s2i run for pytorch model

ukclivecox committed:
Fix formatting in Seldon README to add PyTorch (#95)

* Fix typo in TFMnist URL link

* Update Seldon README to include pyTorch example

kant committed:
Duplicate word on string #15 (#96)

ckadner committed:
Jupyter notebook using Adversarial Robustness Toolbox (#93)

This notebook shows how to integrate the Fast Gradient Method (FGM) from
the Adversarial Robustness Toolbox (ART) into a model training pipeline.
The generated adversarial samples are used to evaluate the robustness of
the trained model. This notebook trains a CNN model on the MNIST hand-
written digit data set and evaluates its robustness. Learn more about
ART at https://github.com/IBM/adversarial-robustness-toolbox

The notebook is adapted for use with FfDL from it original source at
https://github.ibm.com/robust-dlaas/ART-in-WML

animeshsingh committed:
Creating a single place for FfDL collateral for demos, talks, articles, videos etc (#99)

* demo branch

* Update README.md

* Update README.md

* Update README.md

Tomcli committed:
Add instructions for H2O example  (#102)

* add detailed H2O instructions

* add detailed H2O instructions

animeshsingh committed:
H2O arch diagram (#103)

* h20 arch image

* h20 arch image

* h20 arch image

wwalisa committed:
add chinese readme file (#100)

* upload chinese readme file .

upload chinese readme file .

* add chinese readme file hyperlinks on readme file.

add chinese readme file hyperlinks on readme file.

* add chinese readme file hyperlinks on readme file.

add chinese readme file hyperlinks on readme file.

* modify

modify

fplk committed:
Merge 20180514 1536 (#79)

Pre-0.1 release: Add Object Storage mount and other enhancements.

animeshsingh committed:
Arch diag + other minor additions (#108)

* Architecture Details

* Architecture Details

* Architecture Details

Tomcli committed:
Enable Horovod with TensorFlow and PyTorch (#104)

* init commit for horovod patch

* update examples and docs

* update docs and converter script

* update example readme

* update example readme

* modify horovod examples with real workload

* modify horovod examples

* update sed syntax to be more visual friendly

* add troubleshooting for dind cluster

* remove deprecated instructions

animeshsingh committed:
Horovod - Adding details (#109)

* horovod

* horovod

* horovod

* horovod

* horovod

* horovod

* horovod

ckadner committed:
Remove 4 minute timeout for log follow process (#107)

* Remove 4 minute timeout for log follow process (#106)

The process that follows the training logs of an ongoing training job
should not timeout after 4 minutes. Instead the log follow process
should complete after the training job itself is finished.

This behavior is necessary to enable chaining up commands to create
machine learning pipelines, where subsequent commands require the output
data of the training job whose logs are being "followed" like in our
ART notebook.

This commit reinstates the log follow behavior prior merge of PR #79

* Updates suggested by sboagibm

Intention was to not rely on a long term stream being held open, 
but to be able to re-open a new stream starting from where the 
old left off, if the connection terminates.

ckadner committed:
Update ART notebook(s) after PR #79 (#116)

* Update ART Notebook after PR #79

- Load cluster configuration from environment variables
- Require PUBLIC_IP and KUBECONFIG instead of CLUSTER_NAME and VM_TYPE
- Use storage type "mount_cos" (s3fs) instead of "s3_datastore"

* Update ART demo notebook after PR #79

- Load cluster configuration from environment variables
- Require PUBLIC_IP and KUBECONFIG instead of CLUSTER_NAME and VM_TYPE
- Use storage type "mount_cos" (s3fs) instead of "s3_datastore"

ukclivecox committed:
updates for seldon-core 0.2 (#118)

Tomcli committed:
Update deprecated framework versions (#119)

* update dl framework versions

* update examples with new framework tags

Tomcli committed:
Update Fashion MNIST example with Seldon-core 0.2 (#120)

* update fashion mnist example with seldon 0.2

* fix readme

sboagibm committed:
Sboag  dlaas pull june14 (#97)

* Pointed travis testing to do hostmount minikube

* Debugging permissions error.

* Fix to mkdir problems.

* Fixed Makefile syntax.

* Printing debugging information about pods.

* Printing debugging information about pods.

* Printing debugging information about pods.

* Printing debugging information incl kubectl get pod.

* Enabled debug mode.

* Again.

* Set debug as default.

* tracing from the trainer to lcm

* more debugging

* added lower level logging

* dist: xenial

* Update .travis.yml

* fix typo

* Trying to fix Travis issue.

* Fixed Travis issue.

* Followed Tommy's request and increased resource limits to values from before. Might break CI.

* Parameterized memory values like Tommy requested.

* Attempt to fix CI.

* Removed excessive debug statements and cleaned comments. Probably breaks code.

* DLaaS pull june 14, with security mods

* fixed glide problem

* Added Image.go etc. files, deleted learner_test.go

* temporarily disable framework validation

* FIXME: Disable validation check for bucket until conditionalize for s3fs vs.  option.

* fixed two bugs related to volume mounting

* I think mostly just logging changes

* basic success

* Add FfDL.iml to .gitignore

* removed docker ref to csf_env.properties

* Test for mount_cos before attempting s3 validation

* fixed hostmount by pre-setup of model code in Makefile

* fixed missing import

* log HELM_DEPLOY_DIR, add a bunch of logging for the ci test

* Added create-volumes to jenkins file, more verbose docker build for ui

* Wound back Angular to 6.0.8

* Quiet docker-build-ui docker build

* merged bin/create_static_volumes_config2.sh into bin/create_static_volumes_config.sh

Tomcli committed:
Version control for prebuilt images  (#122)

* update prebuild image version, update helm chart to 0.1.1

* fix make deploy bug

Tomcli committed:
Make helm charts and scripts compatible to deploy FfDL on any namespace (#110)

* make helm charts and scripts compatible to deploy FfDL on any namespace

* allow users to export all the enviornment variables in a txt file

* Update readme with new notice

* Fix typo

* Update static volumes config v2 namespace parameter

* capitalize NAMESPACE, update Makefile, developer guide, and trobleshooting.

Tomcli committed:
Fix NAMESPACE typo on one of the deployment script (#126)

Tomcli committed:
use python zipfile instead of bash unzip (#128)

ckadner committed:
ART notebook did not load due to missing comma (#129)

Tomcli committed:
update statsd flags with double dashes (#132)

LGTM. Ran fine / fixed statsd issue on Ubuntu 18.04 Vagrant VM.

animeshsingh committed:
Simplifying README layout (#135)

* Simplifying README

* Simplifying README

* Create detailed-installation-instructions.md

* Update README.md

* Update detailed-installation-instructions.md

* Update and rename detailed-installation-instructions.md to detailed-installation-guide.md

* Update detailed-installation-guide.md

* Update detailed-installation-guide.md

* Update README.md

* Update README.md

* Update README.md

* Update README.md

* Update README.md

* Update README.md

sdmonov committed:
Add ppc64le & multiarch support

This code is adding ppc64le and multiarch support for FfDL.

- Adds new make target to the main Makefile: **docker-create-manifest**
  that creates the multiarch manifests for the services.

- arch (architecture) parameter was added to values.yaml and
  storage-plugin/values.yaml

**make docker-build** will generate all the services with a -${ARCHITECTURE}
suffix. During build an ARCH argument is sent to the Dockerfiles in order
to have architecture specific implementation inside it.

Building docker images should look like this:

make docker-build
make docker-push
make docker-create-manifest

sdmonov committed:
Merge branch 'master' into add_ppc64le_multiarch_support

sdmonov commented:
Hi All,

This code adds support for ppc64le in FfDL. It does adds a suffix to the built services with the architecture and also adds a target in the main Makefile to create/ammend the needed manifest lists to support the multiple architectures.

There are two services that are needed by the code prom/pushgateway and localstack/localstack (optional) that does not have a ppc64le version in the public registries. I did generate custom build ppc64le version and pushed it to my own registry smonov. They will be pulled from there for now until published on a more official registry.

Tomcli commented on values.yaml:
```diff
@@ -2,10 +2,12 @@ namespace: default
   expose_node_port: true
 docker:
-  registry: docker.io
+  registry: ffdl.ibm.com
```
This changes our official Docker image location.

sdmonov replied:
Sorry my mistake. This was not supposed to be there. Will fix it.

sdmonov commented:
Found few issues. Fixing and testing them now:

- fixed issue with the repository in the values.yaml reported by @Tomcli 
- added new target to the Makefile: **docker-tag-local**. It will tag the arch dependent image name to arch independent name. For example: **ffdl/ffdl-ui-ppc64le:v0.1.1** will be tagged to **ffdl/ffdl-ui:v.0.1.1**. This is used for local deployments only that do not require pushing to registry. **make docker-push** will push only the arch dependent images (with -${ARCHITECTURE} suffix) only and then a manifest lists will be created for these image to support multiarch (by invoking make docker-create-manifest).
- updated .travis.yml to invoke docker-tag-local.

animeshsingh commented:
Thanks @sdmonov. What is the target for travis if any?

seelam commented:
>helm install storage-plugin --set arch=ppc64le,namespace=$NAMESPACE
What if you have a mixed x86, POWER cluster, what happens then?

Could we not put the multi-arch storage plugin in a private repo and use it?

The public travis ci is now multi-arch so you should do builds with it. I would like to see the changes in Travis.yaml file. 

We have to deal with learner images... is that addressed in a different PR?

sdmonov committed:
Fixed few issues with ppc64le/multiarch support

- added docker-tag-local target in Makefile
- fixed few issues in docker-create-manifest target in Makefile
- small fix in docker-push to not have duplicate code

sdmonov commented:
@seelam:
- About the storage plugin I totally agree with you we need to push multi-arch images and create a manifest list. I have no access to push images at the moment. If I can be given access (@animeshsingh) I can push both ppc64le and x86_64 images and create the manifest lists to support the multi-arch.
- at the moment travil.yml (as far as I can tell looking into it) is doing testing only. It is not pushing images to the repos
- learner images will go separate PR. will commit soon.
